README.md
![open source](https://badgen.net/badge/open/source/blue?icon=github)

### Disclaimer
This library is **NOT OFFICIALLY SUPPORTED OR ENDORSED BY x-io TECHNOLOGIES**. 
This is made in the hopes of making integration with the x-IMU3 GUI easier for hobbyist projects. 
Both x-io Technologies and the author of the library do not offer any warranty on this library and therefore do not accept liability for its use. 
If you need support on a specific project, please create an issue or a pull request and it will be addressed.

# xioAPI-Arduino

An Arduino Library to facilitate serial communication with the xIMU3 GUI data visualisation software. 
Originally forked from the [ReefWing xIMU3 library]([url](https://github.com/Reefwing-Software/Reefwing-xIMU3)), this library has since expanded to be a general implementation of the xioAPI as specified in the [x-IMU3 user manual]([url](https://x-io.co.uk/downloads/x-IMU3-User-Manual-v1.1.pdf)).

## x-io Technologies & the x-IMU3 GUI

[x-io Technologies](https://x-io.co.uk/) Limited is a UK-based engineering company that specialises in the design and production of IMUs. 
It was founded in 2012 by Seb Madgwick during his PhD at the University of Bristol. 
The [Madgwick algorithm](https://github.com/xioTechnologies/Fusion) in its various incarnations, is used in a LOT of modern AHRS libraries.

In addition to their sensor fusion software, x-io also have an IMU data visualisation tool called the x-IMU3 GUI. 
This free open source software is available for all major operating systems (Windows, macOS, and Linux). 
You can download a copy from the [x-io web site](https://x-io.co.uk/x-imu3/). 

## The x-IMU3 API

The API largely consists of command and data messages. 
An excerpt from the x-IMUs user manual follows:

```
The communication protocol consists of two message types:

- Command messages 
- Data messages 

All messages are terminated by a Line Feed (LF) control character. 
This termination byte will not appear anywhere else in a message and 
so can be used to divide a byte stream into individual messages. 

Some messages are terminated with an additional Carriage Return (CR) 
control character.

The first byte of a message indicates the message type. Command messages start with the character "{" (0x7B in hex, 123 in decimal). 
Data messages start with either an uppercase character or a byte value greater than 0x80 (128 in decimal) depending on the message.
You could use the command messages to control your Arduino from the x-IMU3 GUI, and we will demonstrate how to read and parse the commands from the serial port, but first let's look at data messages.
```

This library makes it possible to integrate Arduino projects directly with the xioAPI and make configuration and visualization easier for the hobbyist programmer.

## Data Messages

One of the key things that you will note when looking at the data message format (Figure 1), is that they all include a time stamp (in microseconds). 
This means that you need an IMU library which returns a time stamp with each sensor reading. 
Some IMUs have a dedicated time stamp register, but others do not. 
Therefore, we must use the Arduino `micros()` function to record the time at which the reading is taken.

| MSG ID | TIME STAMP (µs) |         PAYLOAD         | STOP BYTE |
|:------:|:---------------:|:-----------------------:|:---------:|

*Figure 1. x-IMU3 Data Message Format*

Data messages can be either ASCII encoded or binary, depending on the device settings. 
For now, the only supported message format is ASCII. Each value in an ASCII data message is comma separated.

The `MSG_ID` is the first byte in a data message and needs be a single uppercase character indicating the message type (e.g., `I`). 
The next value is the time stamp in microseconds. 
The API specifies a 64-bit unsigned integer for the time stamp, but 32-bits works, and this is what `micros()` returns.

The data payload data type depends on the message. 
For example, the magnetometer message (`MSG_ID = M`) expects three floating point numbers printed to four decimal places. 
The values expected from the IMU are the scaled output (i.e., celcius for temperature, DPS for gyroscope, g for acceleration, and gauss for the magnetometer). 
All messages are terminated by a Line Feed (`\n`) control character. 
This termination byte will not appear anywhere else in a message and so it can be used to divide a byte stream into individual messages.
Most messages are also terminated with an additional Carriage Return (`\r`) control character.

This library includes wrapper messages for all of the data message types. 
For example, to send a magnetometer message you can use the method, `sendMag(MagnetoMessage msg)`. 
The `MagnetoMessage` struct contains all of the information required to properly format this message. 
It is up to you to build the appropriate pipeline from your sensor readings to the message struct.

``` c++
struct MagnetoMessage {
    float mx, my, mz;
    uint32_t  timestamp;
};
```

All of the possible data messages are currently supported.

## Network Announcement Message

If your device is WiFi-capable, then you can connect it wirelessly to the x-IMU3 GUI over a local network connection.
Your device will need to broadcast a Network Announcement Message every second from UDP port 10000.
The Network Announcement Message is a JSON packet specified below and can be called using the library using the `sendNetworkAnnouncement(NetworkAnnouncement na)` function.
You will need to populate the fields of the NetworkAnnouncement structure based upon your device.

``` JSON
{
  "sync": 0,
  "name": "Raven",
  "sn": "0123-4567-89AB-CDEF",
  "ip": "192.168.1.1",
  "port": 7000,
  "send": 8000,
  "receive": 9000,
  "rssi": 100,
  "battery": 100,
  "status": 0
}
```

## Command Messages

Command messages are a JSON object containing a single key/value pair, terminated by the usual `\r\n`.
A sample command message, requesting the device serial number looks like:

``` JSON
{"serialNumber":null}\r\n
```

The library comes with a function, `checkForCommand()` that polls the API's stream interface for any data.
If there is data present, then the function will read the data until the null terminator ('\n').
The read data will then be attempted to format into a JSON buffer using the [ArduinoJson libary]([url](https://github.com/bblanchon/ArduinoJson)).
If the data is not a valid JSON key/value pair, then an error will be returned and the function will continue checking for commands.
But, if there is a JSON string present, it will parse it into separate key and value variables for later use.
The function will also immediately handle the command using the `handleCommand()` function.

## Parsing Command Messages

Since there are many possible command messages that can be sent to the device, it is necessary to handle them in special ways to save memory at the cost of increased processing requirements.
The command value will always be a string and therefore difficult to process in Arduino.
An easy solution is to assign this string a specific numerical ID that we can quickly compare using a `switch-case` statement.
Therefore, this library uses the [DJB2 hash function]([url](https://theartincode.stanis.me/008-djb2/)) to encode the command string into an `unsigned long`.
One thing to note is that the output of the hashing function is dependent on the architecture.
On most Arduino-compatible microcontrollers, the `unsigned long` data type represents 32-bits.
However, on others it may be 64-bits.
If you find that the library is throwing an error about invalid keys, this may be the problem.

## Device Settings

The device settings are stored in a structure called `settings`.
To access them at any time you may call: `settings.accelerometerOffset`, for instance.
Settings can be loaded or saved to a JSON file stored in an onboard filesystem using the `loadConfigurationsFromJSON()` function.
Note that for now, the only supported filesystem for this feature is SPIFFS, commonly used on the ESP32 platforms (this is being addressed in [Issue #2]([url](https://github.com/Legohead259/xioAPI-Arduino/issues/2)))

To change settings, it is recommended to use the `updateSetting()` function, though for now, you will need to pass in the new value manually as a `JsonVariant` object.
This will prevent unintended behavior from affecting the rest of the system.

## Connecting x-IMU3 GUI to your Arduino

Plug you Arduino into a USB port on your PC and work out the name of the port it is connected to. You can see this in the Arduino IDE. Then open up the x-IMU3 GUI app, click on `Connection` (top left) -> `New USB Connection`. In the dialogue box that pops up, select the port with the Arduino connect and click on Connect.

## Using the xioAPI-Arduino Library


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!-- 
Release sections
## [TAG] Version X.X.X - YYYY-MM-DD
### Known Bugs
### Added
### Changed
### Fixed
### Deprecated
### Removed
### Security 
-->

## [UNRELEASED] Version 1.1.0 - 2023-06
### Known Bugs
- xioAPI UDP functionality not working as intended. [#6](https://github.com/Legohead259/xioAPI-Arduino/issues/6)
  
### Added
- Added new `updateSetting()` functions that can access the setting lookup table directly
- Added new `xioMatrix` and `xioVector` types
- Added support for UDP WiFi messaging
- Added dependency for `WiFiUDP`
- Added a CircularBuffer to hold data for the datalogger

### Changed
- Minor refactor of `sendTime()` to `cmdReadTime()` for clarity and consistency
- Changed type of `displayName`, `ipAddress`, and `serialNumber` to character array from character pointer
- Changed `send()` to better generalize support between different interfaces

### Removed
- Removed `print()` functionality
- Removed `writeLenFeed` arguments since all calls require a linefeed
  
---

## Version 1.0.0 - 2023-06-16
### Known Bugs
- Null Pointer error on some commands
- Configuration storage only compatible with JSON in SPIFFS [#2]([url](https://github.com/Legohead259/xioAPI-Arduino/issues/2))
- ESP32-S2 does not connect to x-IMU3 GUI natively, despite the functions working as intended [#3]([url](https://github.com/Legohead259/xioAPI-Arduino/issues/3))

### Added
- Basic implementation of all xioAPI commands
- Configuration settings and handling
- Added a `readAll` command the sends all settings from the settings lookup table
- Added a `readJson` command the sends the JSON configuration setting file

src/xioAPI.cpp
/******************************************************************
    @file       xioAPI.cpp
    @brief      Arduino-compatible implementation of the xioAPI for embedded systems
    @author     Braidan Duffy
    @copyright  MIT License

    Code:       Braidan Duffy
    Version:    1.0.0
    Date:       18/05/2023
    Modified:   18/05/2023

    CHANGELOG:
    v1.0.0 - Original release
    
    Credit - Derived from the xIMU3 User Manual 
            (https://x-io.co.uk/downloads/x-IMU3-User-Manual-v1.1.pdf).
******************************************************************/

#include "xioAPI.h"

xioAPI api;
CircularBuffer<char,8192> dataASCIIBuffer;

using namespace xioAPI_Types;
using namespace xioAPI_Protocol;

/**
 * @brief Initializes the API for communication
 * 
 * @param port The Stream interface to be used for communication
*/
bool xioAPI::begin(Stream* port) {
    _serialPort = port;
    _isActive = true;
    return true;
}

bool xioAPI::begin(Stream* port, WiFiUDP* server) {
    _udpServer = server;
    return begin(port);
}


// =========================
// === COMMAND FUNCTIONS ===
// =========================


/**
 * @brief Sends a device setting to the requestor
 * 
 * @param key The setting name (consult the xioAPI documentation for valid names)
 * @param entry The corresponding entry from the settings lookup table
*/
void xioAPI::sendSetting(const settingTableEntry* entry) {
    // Setting format: {"[setting]":[value]}\r\n

    StaticJsonDocument<256> _doc;
    char _out[256];

    const char* key = entry->key;
    SettingType _entryType = entry->type;
    void* _entryValue = entry->value;
    JsonArray _settingArray;
    xioVector* _vector;
    xioMatrix* _matrix;

    switch (_entryType) {
        case BOOL:
            _doc[key] = *(bool*) _entryValue;
            break;
        case CHAR:
            _doc[key] = *(uint8_t*) _entryValue;
            break;
        case FLOAT:
            _doc[key] = *(float*) _entryValue;
            break;
        case INT:
            _doc[key] = *(int*) _entryValue;
            break;
        case VECTOR:
            _settingArray = _doc.createNestedArray(key);
            _vector = static_cast<xioVector*>(_entryValue);
            for (size_t i=0; i<3; i++) {
                _settingArray.add(_vector->array[i]);
            }
            break;
        case MATRIX:
            _settingArray = _doc.createNestedArray(key);
            _matrix = static_cast<xioMatrix*>(_entryValue);
            for (size_t i=0; i<9; i++) {
                _settingArray.add(_matrix->array[i/3][i%3]);
            }
            break;
        case CHAR_ARRAY:
            _doc[key] = (const char*) _entryValue;
            break;
        default:
            return;
    }

    size_t outLen = serializeJson(_doc, _out);
    send("%s", _out);
}

/**
 * @brief Sends a ping to the requestor containing device information.
 * 
 * @param ping A Ping structure containing the currently used interface, the device name, and the device serial number
*/
void xioAPI::sendPing(Ping ping) {
    /**
     * Ping format:
     * {
     * "ping": {
     *      "interface": "[interface]",
     *      "deviceName": "[deviceName]",
     *      "serialNumber": "[serialNumber]"
     * }
     * }\r\n
    */
    StaticJsonDocument<96> _doc;
    char _out[96];

    JsonObject root = _doc.createNestedObject("ping");
    root["interface"] = "USB";
    root["deviceName"] = "Thetis";
    root["serialNumber"] = "Unknown";

    size_t outLen = serializeJson(_doc, _out);
    send("%s", _out);
}

/**
 * @brief The network announcement message is used by a host to discover and connect to devices on the same network.
 * The message is continuously broadcast by the device on UDP port 10000 at 1 Hz.
 * 
 * @param na A NetworkAnnouncement structure containing the device sync word, name, serial number, IP address, TCP port, UDP send port, UDP receive port, network RSSI, battery percentage, and charging status.
*/
void xioAPI::sendNetworkAnnouncement(NetworkAnnouncement na) {
    /**
     * Network announcement format:
     * {
     * "sync": [sync],
     * "name": "[displayName]",
     * "sn": "[serialNumber]",
     * "ip": "[ipAddress]",
     * "port": [tcpPort],
     * "send": [udpSendPort],
     * "receive": [udpReceivePort],
     * "rssi": [wiFiRSSI]
     * "battery": [batteryPercentage],
     * "status": [chargingStatus]
     * }\r\n
    */
    StaticJsonDocument<256> _doc;
    char _out[256];

    _doc["sync"] = na.sync;
    _doc["name"] = na.displayName;
    _doc["sn"] = na.serialNumber;
    _doc["ip"] = na.ipAddress;
    _doc["port"] = na.tcpPort;
    _doc["send"] = na.udpSendPort;
    _doc["receive"] = na.udpReceivePort;
    _doc["rssi"] = na.rssiPercentage;
    _doc["battery"] = na.batteryPercentage;
    _doc["status"] = na.chargingStatus;

    size_t outLen = serializeJson(_doc, _out);
    sendUDP((uint8_t*) _out, outLen, "255.255.255.255", 10000);
}

void xioAPI::sendSettingTable() {
    for (size_t i=0; i<SETTING_TABLE_SIZE; i++) {
        if (settingTable[i].key == nullptr) continue; // Skip sending the setting if the key (entry) is empty
        sendSetting(&settingTable[i]);
    }
}

void xioAPI::sendSettingFile() {
    char _out[6144];
    size_t outLen = serializeJsonPretty(_jsonConfigDoc, _out);
    send("%s", _out);
}


// =================================
// === COMMAND HANDLER FUNCTIONS ===
// =================================


/**
 * @brief Continuously checks if a command is available from the device interface.
 * If a command is present, it will parse it into a command key and a value for later processing.
 * The command will automatically be processed as soon as it is detected.
*/
void xioAPI::checkForCommand() {
    char buffer[256];
    StaticJsonDocument<256> doc;

    while (_serialPort->available() > 0) { //  Check for xio API Command Messages
        // Read the incoming bytes
        int blen = _serialPort->readBytesUntil(TERMINAL, buffer, BUFFER_SIZE);
        DeserializationError error = deserializeJson(doc, buffer, blen);

        if (error) {
            _serialPort->println("Ran into error"); // DEBUG
            _serialPort->println(error.c_str()); // DEBUG
            return;
        }

        JsonObject root = doc.as<JsonObject>();

        for (JsonPair kv : root) {
            strcpy(_cmd, kv.key().c_str());
            _value = kv.value();
        }

        handleCommand(_cmd);
    }
}

/**
 * @brief Handles the command based upon the key presented. 
*/
void xioAPI::handleCommand(const char* cmdPtr) {
    uint32_t cmdHash = hash(cmdPtr);

    using xioAPI_Protocol::APIKeyHashASCII;

    // First check to see if it is a setting read/write command:
    const size_t numEntries = sizeof(settingTable) / sizeof(settingTable[0]);
    for (size_t i=0; i<numEntries; i++) { // Check all keys in the hash table
        if (settingTable[i].hash == cmdHash) { // If the command key is present in the hash table, then it is a setting read/write
            if (_value.isNull()) { // If the passed value was null, then it is a read command
                sendSetting(&settingTable[i]);
            }
            else {
                updateSetting(&settingTable[i], _value);
                sendSetting(&settingTable[i]);
            }
            
            // Exit function after handle
            return;
        }
    }

    switch(cmdHash) {
        case XIO_DEFAULT:
            loadConfigurationsFromJSON(true, DEFAULT_CONFIG_FILE_NAME);
            break;
        case APPLY:
            // TODO: ignore? Settings are automatically set when configurations loaded
            sendAck("apply");
            break;
        case SAVE:
            saveConfigurations();
            sendAck("save");
            break;
        case TIME:
            if (_value != nullptr) { // Assume that a WRITE command has been sent
                cmdWriteTime();
            }
            cmdReadTime();
            break;
        case PING:
            sendPing(Ping{"USB", settings.deviceName, settings.serialNumber}); // TODO: Update to support multiple interfaces
            break;
        case RESET:
            sendAck("reset");
            cmdReset();
            break;
        case SHUTDOWN:
            sendAck("shutdown");
            cmdShutdown();
            break;
        case STROBE:
            cmdStrobe();
            break;
        case COLOUR:
            cmdColour();
            break;
        case HEADING:
            if (!settings.ahrsIgnoreMagnetometer) {
                cmdHeading();
            }
            send("{\"heading\":%0.3f}", getValue<float>());
            break;
        case xioAPI_Protocol::ACCESSORY:
            cmdSerialAccessory();
            sendAck("serialAccessory");
            break;
        case NOTE:
            sendNotification(getValue<const char*>());
            break;
        case FORMAT:
            cmdFormat();
            sendAck("format");
            break;
        case TEST:
            cmdSelfTest();
            break;
        case BOOTLOADER:
            sendAck("bootloader");
            cmdBootloader();
            break;
        case FACTORY:
            cmdFactory();
            sendAck("factory");
            break;
        case ERASE:
            cmdErase();
            sendAck("erase");
            break;
        case READ_ALL:
            sendSettingTable();
            break;
        case READ_JSON:
            sendSettingFile();
            break;
        default:
            char _buf[128];
            sprintf(_buf, "Did not recognize key: %s as %08x", cmdPtr, cmdHash);
            sendError(_buf);
    }
}


// ====================
// === API MESSAGES ===
// ====================


void xioAPI::sendInertialMessage(InertialMessage msg) {
    // Inertial Message Format: "I,timestamp (µs),gx,gy,gz,ax,ay,az\r\n"
    sendDataMessage("I,%lu,%0.4f,%0.4f,%0.4f,%0.4f,%0.4f,%0.4f", msg.timestamp, msg.gx, msg.gy, msg.gz, msg.ax, msg.ay, msg.az);
}

void xioAPI::sendMagnetometerMessage(MagnetometerMessage msg) {
    // Magnetometer Message Format: "M,timestamp (µs),mx,my,mz\r\n"
    sendDataMessage("M,%lu,%0.4f,%0.4f,%0.4f", msg.timestamp, msg.mx, msg.my, msg.mz);
}

void xioAPI::sendTemperatureMessage(TemperatureMessage msg) {
    // Temperature Message Format: "T,timestamp (µs),temperature (°C)\r\n"
    sendDataMessage("T,%lu,%0.4f", msg.timestamp, msg.temp);
}

void xioAPI::sendQuaternionMessage(QuaternionMessage msg) {
    // Quaternion Message Format: "Q,timestamp (µs),w,x,y,z\r\n"
    sendDataMessage("Q,%lu,%0.4f,%0.4f,%0.4f,%0.4f", msg.timestamp, msg.w, msg.x, msg.y, msg.z);
}

void xioAPI::sendEulerMessage(EulerMessage msg) {
    // Euler Angles Message Format: "A,timestamp (µs),roll,pitch,yaw\r\n"
    sendDataMessage("A,%lu,%0.4f,%0.4f,%0.4f", msg.timestamp, msg.roll, msg.pitch, msg.yaw);
}

void xioAPI::sendBatteryMessage(BatteryMessage msg) {
    // Battery Message Format: "B,timestamp (µs),percentCharged,voltage,status\r\n"
    sendDataMessage("B,%lu,%0.4f,%0.4f,%u", msg.timestamp, msg.percentCharged, msg.voltage, msg.status);
}

void xioAPI::sendRSSIMessage(RSSIMessage msg) {
    // RSSI Message Format: "W,timestamp (µs),percent,power (dBm)\r\n"
    sendDataMessage("W,%lu,%0.4f,%0.4f", msg.timestamp, msg.percentage, msg.power);
}

void xioAPI::sendNotification(const char *note) {
    // Notification Message Format: "N,timestamp (µs),note\r\n"
    send("N,%lu,%s", micros(), note);
}

void xioAPI::sendError(const char *error) {
    // Error Message Format: "F,timestamp (µs),errorMessage\r\n"
    send("F,%lu,%s", micros(), error);
}


// =========================
// === UTILITY FUNCTIONS ===
// =========================


void xioAPI::send(const char* message, ...) {
    char buffer[128];
    va_list args;
    va_start(args, message);
    size_t writeLen = vsnprintf(buffer, sizeof(buffer), message, args);
    va_end(args);

    sendSerial(buffer, writeLen);
    
    sendUDP((uint8_t*) buffer, writeLen);
}

void xioAPI::sendDataMessage(const char* message, ...) {
    char buffer[128];
    va_list args;
    va_start(args, message);
    size_t writeLen = vsnprintf(buffer, sizeof(buffer), message, args);
    va_end(args);

    if (settings.usbDataMessagesEnabled) sendSerial(buffer, writeLen);
    if (settings.udpDataMessagesEnabled) sendUDP((uint8_t*) buffer, writeLen);

    if (settings.dataLoggerDataMessagesEnabled) {
        for (int i=0; i<writeLen; i++) {
            dataASCIIBuffer.push(buffer[i]);
        }
        dataASCIIBuffer.push('\n');
    }
}

void xioAPI::sendSerial(const char* buffer, size_t size) {
    if (_serialPort != nullptr) {
        if (settings.binaryModeEnabled) {
            if (_serialPort != nullptr) { // Write data to serial terminal, if available.
                _serialPort->write((uint8_t*) buffer, size);
                _serialPort->write(0x0D);
                _serialPort->write(0x0A);
            }
        }
        else {
            _serialPort->print(buffer);
            _serialPort->print("\r\n");
        }
    }
}

void xioAPI::sendUDP(uint8_t* buffer, size_t size, char* ipAddress, int sendPort) {
    if (_udpServer != nullptr && settings.wirelessMode) { // Write data to UDP unicast, if available
        _udpServer->beginPacket(ipAddress, sendPort);
        _udpServer->write(buffer, size);
        _udpServer->write(0x0D);
        _udpServer->write(0x0A);
        _udpServer->endPacket();
    }
}

/**
 * @brief Clears the value of the `_cmd` pointer 
*/
void xioAPI::clearCmd() {
    memset(_cmd, 0, sizeof(_cmd));
}

/**
 * @brief Clears the value of the `_value` object
*/
void xioAPI::clearValue() {
    _value.clear();
}

/**
 * @deprecated
 * @brief Determines what the type of a given character is based on the JSON standard.
 * 
 * @param c The character to check
 * 
 * @return the type of the given character in terms of the JSON standard
*/
ValueType xioAPI::parseValueType(char c) {
    switch(c) {
        case 'n':
            return JSON_NULL;
            break;
        case 't':
        case 'f':
            return JSON_BOOL;
            break;
        case '-': case '0': case '1': case '2': case '3': case '4': 
        case '5': case '6': case '7': case '8': case '9':
            return JSON_NUMBER;
            break;
        case '\"':
            return JSON_STRING;
            break;
        default:
            return JSON_UNDEFINED;
            break;
    }
}